### ISSUE TITLE
Accuracy is too high for inputs that are predicted wrong.

### ISSUE THREAD

**Issue by igorvishnevskiy** [tflearn/tflearn#893]
I have created an NLP LSTM RNN model using following example:
https://github.com/tflearn/tflearn/blob/master/examples/nlp/lstm.py

Model has trained and performs great predicting inputs that exist in the training set or close to those in training set. It predicts them very accurately actually.

**>> The problem** however is that I always receive one high accuracy rate for those inputs that have no related examples provided for in the training set. Below I specified the accuracies that I receive for different inputs. As you can see, those that are predicted correctly, have slightly different accuracy rates. However those that have not been predicted correctly, always have one accuracy rate that is set to the same value for all such inputs: "0.68496119976043701172". That value is quite high. And since I choose the highest accuracy result, I would always get this incorrect prediction with high accuracy that is above my threshold. When I trained with more epochs I would get that accuracy rate raise even higher to 84% about. What am I not doing correctly? For the inputs that do not have any similar examples provided for in the training set, I should receive much lower accuracy rates, not what I receive. I would expect it to predict something, but should not be with accuracy of 68% or higher. S.O.S. I need help. Thank you so much for helping me!

## **Code of my network:**

    net = tflearn.input_data([None, len(train_x[0])])
    net = tflearn.embedding(net, input_dim=10000, output_dim=128)
    net = tflearn.lstm(net, 128, dropout=0.8)
    net = tflearn.fully_connected(net, len(train_y[0]), activation='softmax')
    net = tflearn.regression(net, optimizer='adam', learning_rate=0.001, loss='categorical_crossentropy')
    model = tflearn.DNN(net, tensorboard_dir='tflearn_logs')
    model.fit(train_x, train_y, n_epoch = 1000, shuffle=True, validation_set=(train_x, train_y), show_metric=True, batch_size=64)

## **Examples of too high of the accuracy for incorrect predictions:**

_INCORRECT PREDICTION  - Input has nothing similar to it in the training set that I used for this training_

Input:
ajhhjasvdhgs823754shg83w74 asjhasj nope come on never

sentence_words
['ajhhjasvdhgs823754shg83w74', 'asjhasj', 'nope']

0.00035618932452052832
0.00002501290146028623
0.00044306437484920025
0.00509987585246562958
0.00013837174628861248
0.00008657085709273815
0.00148472678847610950
0.00004535137850325555
0.00080926803639158607
0.00313436333090066910
0.00195945147424936295
0.00264072138816118240
0.00138593628071248531
0.00004885421367362142
0.68496119976043701172
0.00038353921263478696
0.00009397857502335683
0.00078909739386290312
0.00368964602239429951
0.00059157644864171743
0.00353096914477646351
0.00132497865706682205
0.00003743696652236395
0.00140456133522093296
0.00009372259955853224
0.00089485925855115056
0.25274342298507690430
0.00002111023241013754
0.00984090939164161682
0.00004694214658229612
0.02175975404679775238
0.00013464479707181454

{'confidence': 0.6849612, 'intent': 'get_company_contact_information', 'response': u"- company name - contact information can be located on the following page of our website: URL"}

**+++++++++++++++++++++**

_INCORRECT PREDICTION - Input has nothing similar to it in the training set that I used for this training_

Input:
Flying cars for mankind

sentence_words
['flying', u'car', 'mankind']

0.00035618932452052832
0.00002501290146028623
0.00044306437484920025
0.00509987585246562958
0.00013837174628861248
0.00008657085709273815
0.00148472678847610950
0.00004535137850325555
0.00080926803639158607
0.00313436333090066910
0.00195945147424936295
0.00264072138816118240
0.00138593628071248531
0.00004885421367362142
0.68496119976043701172
0.00038353921263478696
0.00009397857502335683
0.00078909739386290312
0.00368964602239429951
0.00059157644864171743
0.00353096914477646351
0.00132497865706682205
0.00003743696652236395
0.00140456133522093296
0.00009372259955853224
0.00089485925855115056
0.25274342298507690430
0.00002111023241013754
0.00984090939164161682
0.00004694214658229612
0.02175975404679775238
0.00013464479707181454

{'confidence': 0.6849612, 'intent': 'get_company_contact_information', 'response': u"- company name - contact information can be located on the following page of our website: URL"}

**+++++++++++++++++++++**

## **Examples of correct predictions:**

_CORRECT PREDICTION_

Input:
Where can I find your products documentation?

sentence_words
[u'product', 'documentation']

0.00612614396959543228
0.00011750899284379557
0.00069767667446285486
0.00085104035679250956
0.00008564032759750262
0.00028970753191970289
0.00068009214010089636
0.00120029202662408352
0.00062775332480669022
0.00147092458792030811
0.00233637029305100441
0.00895396433770656586
0.00052287138532847166
0.00035432158620096743
0.00243101338855922222
0.00387038756161928177
0.00194357126019895077
0.00049759971443563700
0.89442783594131469727
0.00236937170848250389
0.00370639306493103504
0.00438989093527197838
0.00004709115455625579
0.00104782730340957642
0.00004982457539881580
0.00015152101696003228
0.01096363924443721771
0.00006660592043772340
0.00367720634676516056
0.00037029836676083505
0.04505829513072967529
0.00061725842533633113

{'confidence': 0.89442784, 'intent': 'get_product_documentaton', 'response': u"Please navigate to the following web address for Complete - company name - Product Documentation: URL"}

**+++++++++++++++++++++**

_CORRECT PREDICTION_

Input:
I need to create one more issue ticket

sentence_words
['create', 'one', 'issue', 'ticket']

0.00036033347714692354
0.00085572013631463051
0.00156056694686412811
0.00330810411833226681
0.00003927011493942700
0.00007325618207687512
0.01852107793092727661
0.01555508282035589218
0.00225663976743817329
0.00266261305660009384
0.00165271642617881298
0.00103618251159787178
0.00292959786020219326
0.00010923387162620202
0.00389660405926406384
0.00816982984542846680
0.00009279943333240226
0.00008974337106337771
0.00987286027520895004
0.00082714710151776671
0.00068086449755355716
0.00123500928748399019
0.00066235440317541361
0.00063234707340598106
0.00007008510874584317
0.00058565364452078938
0.00426529068499803543
0.00021767383441329002
0.00110195914749056101
0.00031603770912624896
0.91631835699081420898
0.00004503972741076723

{'confidence': 0.91631836, 'intent': 'submit_product_issue', 'response': u"What - company name - product do you need help with? You can say for example: 'I am having issues with ....'"}

**+++++++++++++++++++++**

_CORRECT PREDICTION_

Input:
I need to create one more issue ticket

sentence_words
['create', 'one', 'issue', 'ticket']

0.00036033347714692354
0.00085572013631463051
0.00156056694686412811
0.00330810411833226681
0.00003927011493942700
0.00007325618207687512
0.01852107793092727661
0.01555508282035589218
0.00225663976743817329
0.00266261305660009384
0.00165271642617881298
0.00103618251159787178
0.00292959786020219326
0.00010923387162620202
0.00389660405926406384
0.00816982984542846680
0.00009279943333240226
0.00008974337106337771
0.00987286027520895004
0.00082714710151776671
0.00068086449755355716
0.00123500928748399019
0.00066235440317541361
0.00063234707340598106
0.00007008510874584317
0.00058565364452078938
0.00426529068499803543
0.00021767383441329002
0.00110195914749056101
0.00031603770912624896
0.91631835699081420898
0.00004503972741076723

{'confidence': 0.91631836, 'intent': 'submit_product_issue', 'response': u"What - company name - product do you need help with? You can say for example: 'I am having issues with ....'"}

**+++++++++++++++++++++**

_CORRECT PREDICTION_

Input:
How are you bot

sentence_words
['you', 'bot']

0.00178559648338705301
0.00133074948098510504
0.00027116999262943864
0.02856334298849105835
0.00220576790161430836
0.00076102046296000481
0.00604180758818984032
0.00046779369586147368
0.00036897676181979477
0.00884445756673812866
0.00867486372590065002
0.00998522993177175522
0.00275757559575140476
0.01562077831476926804
0.00771301519125699997
0.00126908253878355026
0.00027335909544490278
0.00203047576360404491
0.00422610156238079071
0.00466328905895352364
0.00738375540822744370
0.00825943239033222198
0.00026308928499929607
0.00243258429691195488
0.00214211596176028252
0.00018701943918131292
0.01025947183370590210
0.00180692458525300026
0.76838761568069458008
0.00528547680005431175
0.01701507344841957092
0.06872295588254928589

{'confidence': 0.76838762, 'intent': 'greeting_how_are_you', 'response': u"I'm fine, Thank you for asking. I am here to help you find information about - company name - products. Please let me know what you are looking for?"}

**+++++++++++++++++++++**

_CORRECT PREDICTION_

Input:
whats up bot

sentence_words
['up', 'bot']

0.00010749076318461448
0.00032820369233377278
0.01560913864523172379
0.69569045305252075195
0.00057597202248871326
0.00067718361970037222
0.00241628172807395458
0.00179249793291091919
0.00070093647809699178
0.00467830337584018707
0.00495495973154902458
0.00330604217015206814
0.00955209229141473770
0.00068924174411222339
0.00623606564477086067
0.00237366184592247009
0.00020166784815955907
0.00352962105534970760
0.00449099112302064896
0.00038545206189155579
0.00571734830737113953
0.00052963860798627138
0.00179245695471763611
0.02322339266538619995
0.00093399430625140667
0.00093780242605134845
0.16592630743980407715
0.00039016545633785427
0.00223310827277600765
0.00021289504365995526
0.03942415863275527954
0.00038231405778788030

{'confidence': 0.69569045, 'intent': 'funny_greeting', 'response': u'Hi, how are you human?'}

**Comment by igorvishnevskiy** [comment]
If you sum the accuracy scores of all predictions for one given input, it totals to 1.0. So it seems that model predicts all correctly except one, that it simply assigns accuracy by subtracting sum of all other from 1.0, which is totally incorrect.

**Comment by balaram-bhukya** [comment]
which technique did you use to convert  (sentence,Label) into train_x and train_y. what is there in train_x and train_y ??

**Comment by igorvishnevskiy** [comment]
I used Bag of Words technique. Below are the code snippets of code that generates train_x and train_y and further down is the sample of the dataset.

**The code that generates train_x and train_y is below:**
############# start of code ################
for data_point in training_data['data']:
        for pattern in data_point['patterns']:
            # tokenize each word in the sentence
            fuse = 0
            pattern = Utils.clean_evil_chars(pattern)
            intent = Utils.clean_evil_chars(data_point["tag"])
            responses = Utils.clean_evil_chars_in_list(data_point["responses"])

            input = nltk_obj.remove_punctuations_marks(pattern.lower())

            # clean_word_set = set(convert_plurals_to_singular(nltk_obj.remove_stop_words(input)))
            # clean_word_list = nltk_obj.sort_list_alphabetically(list(clean_word_set))

            clean_word_list = convert_plurals_to_singular(nltk_obj.remove_stop_words(input))

            #print clean_word_list

            # add to our words list
            words.extend(clean_word_list)
            # add to documents in our corpus
            response_to_classify = '{"intent":"'+str(intent)+'","responses":'+str(responses)+'}'
            documents.append((clean_word_list, response_to_classify))
            # add to our classes list
            if response_to_classify not in classes:
                classes.append(response_to_classify)

    print("Finished generating dataset")

    words = sorted(list(set(words)))

    # remove duplicates
    classes = sorted(list(set(classes)))

    print("Removed duplicates")
    # print (len(documents), "documents")
    # print (len(classes), "classes", classes)
    # print (len(words), "unique stemmed words", words)

    print (len(documents), "documents")
    print (len(classes), "classes")
    print (len(words), "unique stemmed words")

    # create our training data
    training = []
    output = []
    # create an empty array for our output
    output_empty = [0] * len(classes)

    print("Generating bag of words")
    for doc in documents:
        try:
            # initialize our bag of words
            bag = []
            pattern_words = doc[0]
            doc_one = doc[1]
            # create our bag of words array
            for w in words:
                bag.append(1) if w in pattern_words else bag.append(0)

            # output is a '0' for each tag and '1' for current tag
            output_row = list(output_empty)
            output_row[classes.index(doc_one)] = 1
            training.append([bag, output_row])
        except Exception, e:
            print "Exception inside stemming loop"
            print e
            exc_type, exc_obj, exc_tb = sys.exc_info()
            print "Line number: "+str(exc_tb.tb_lineno)
            pass

    print("Finished generating bag of words")

    print("shuffle our features and turn into np.array")
    random.shuffle(training)
    training = np.array(training)

    print("create train and test lists")
    train_x = list(training[:,0])
    train_y = list(training[:,1])
############# end of code ################


**My training dataset example is (real one has many more entries, below set is short enough to fit into this post. Dataset is in JSON format):**
############# start of code ################
{
  "data": [
    {
      "patterns": [""],
      "tag": "intent_default_no_response",
      "responses": [
        "Could not find answer."
      ],
      "context_set": ""
    },
    {
      "patterns": [
        "whats up",
        "Salut",
        "bonjour",
        "Hi",
        "Greetings!",
        "Good day",
        "Hello",
        "Good morning",
        "Good evening"
      ],
      "tag": "greeting",
      "responses": [
        "Hi there! I am here to help you find information about our products. Please let me know what you are looking for?"
      ],
      "context_set": ""
    },
    {
      "patterns": [
        "What is your contact info",
        "What are your contact details",
        "What is your phone number",
        "What is your contact email",
        "How can we get hold of you",
        "How we can talk to the Tibco operator"
      ],
      "tag": "get_company_contact_information",
      "responses": [
        "Our contact information can be located on the following page of our website: URL"
      ],
      "context_set": ""
    },
    {
      "patterns": [
        "Hi, How is it going?",
        "How do you do?",
        "How are you?",
        "How are you bot?",
        "How are you doing?",
        "How r ya doing",
        "How ya doin'",
        "How r u doing?"
      ],
      "tag": "greeting_how_are_you",
      "responses": [
        "I'm fine, Thank you for asking. I am here to help you find information about our products. Please let me know what you are looking for?"
      ],
      "context_set": ""
    }
  ]
}
############# end of code ################

**Comment by balaram-bhukya** [comment]
but in this technique (BOW)  LSTM remembers the sequence of 1's and 0's and labels for them. aren't you missing sequence of words in your training data ??? 
To avoid random classification for random text you can verify if a word exists in BOW then go for score else Null

**Comment by igorvishnevskiy** [comment]
Thank you @balaram-bhukya. You are saying if I will replace what I have in my "train_x" with vectorized sequences of words, it will improve the result? I can try. Is there any helper function in tflearn that helps preprocess data in that format by any chance? Thank you so much for your help.

**Comment by igorvishnevskiy** [comment]
I ended up rewriting data preprocessing code to vectorize words from the dataset, pad sequences of vectors representing the sequences of words in the given input sentence and then pass those sequences as X into ML network. Indeed model started working much better and started returning lower accuracy for predictions that it could not find a match for. I however had to build a custom lemmatization feature, the modules that I found were stemming words way too much. One example is word "operator" was stemmed into "oper", which would not work for me. Therefore I build my own that converted all verbs into infinitive state and converted all plural nouns into singular. That worked much better. Right now working on data augmentation, if any issues, will submit separate ticket for it. For now, closing this one. 

Thank you @balaram-bhukya for your help.
